node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1289,11 +1295,226 @@ async fn fetch_block_worker(
+        validator_weights: &BTreeMap<PublicKey, U512>,
+        finality_threshold_fraction: Ratio<u64>,
+    ) -> bool {
+        self.0.as_ref().map_or(false, |sigs| {
+            are_signatures_sufficient_for_sync_to_genesis(
+                consensus::check_sufficient_finality_signatures(
+                    validator_weights,
+                    finality_threshold_fraction,
+                    sigs,
+                ),
+            )
+        })
+    }
+
+    fn into_inner(self) -> Option<BlockSignatures> {
+        self.0
+    }
+}
+
+// Returns true if the output from consensus can be interpreted
+// as sufficient finality signatures for the sync to genesis process.
+fn are_signatures_sufficient_for_sync_to_genesis(
+    consensus_verdict: Result<(), FinalitySignatureError>,
+) -> bool {
+    match consensus_verdict {
+        Ok(_) | Err(FinalitySignatureError::TooManySignatures { .. }) => true,
+        Err(_) => false,
+    }
+}
+
+// Fetches the finality signatures from the given peer. In case of timeout, it'll
+// retry up to `retries` times. Other errors interrupt the process immediately.
+async fn fetch_finality_signatures_with_retry(
+    block_hash: BlockHash,
+    peer: NodeId,
+    retries: usize,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<FetchedData<BlockSignatures>, FetcherError<BlockSignatures>> {
+    for _ in 0..retries {
+        let maybe_signatures = ctx
+            .effect_builder
+            .fetch::<BlockSignatures>(block_hash, peer)
+            .await;
+        match maybe_signatures {
+            Ok(result) => return Ok(result),
+            Err(FetcherError::TimedOut { .. }) => continue,
+            Err(_) => return maybe_signatures,
+        }
+    }
+
+    Err(FetcherError::TimedOut {
+        id: block_hash,
+        peer,
+    })
+}
+
+async fn fetch_finality_signatures_by_block_header(
+    block_header: BlockHeader,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<Option<BlockSignatures>, Error> {
+    let start = Timestamp::now();
+    let mut peer_list = ctx
+        .effect_builder
+        .get_fully_connected_non_joiner_peers()
+        .await;
+    ctx.filter_bad_peers(&mut peer_list);
+
+    let mut sig_collector = BlockSignaturesCollector::new();
+
+    let block_header_hash = block_header.hash(ctx.config.verifiable_chunked_hash_activation());
+
+    for peer in peer_list {
+        let fetched_signatures = fetch_finality_signatures_with_retry(
+            block_header_hash,
+            peer,
+            FINALITY_SIGNATURE_FETCH_RETRY_COUNT,
+            ctx,
+        )
+        .await;
+
+        let signatures = match fetched_signatures {
+            Ok(FetchedData::FromStorage { item, .. }) => {
+                trace!(
+                    ?block_header_hash,
+                    ?peer,
+                    "fetched FinalitySignatures from storage",
+                );
+                ctx.metrics
+                    .observe_fetch_finality_signatures_duration_seconds(start);
```

Comment:
Now I think that it might be a good idea to NOT record if we get it from the storage - it will be skewing the metrics towards almost-0 reading.